src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -155,25 +205,195 @@ protected final SCMSourceCriteria getCriteria() {
+     * the event information to check the validity of any intermediary caches..
+     *
+     * @param <O> Observer type.
+     * @param criteria the criteria to use.
+     * @param observer an optional observer of interim results.
+     * @param event the event from which the fetch should be scoped.
+     * @param listener the task listener
+     * @return the provided observer.
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     * @see SCMEvent#isPayloadTrusted()
+     * @since 2.0
+     */
+    @NonNull
+    public final <O extends SCMHeadObserver> O fetch(@CheckForNull SCMSourceCriteria criteria,
+                                                     @NonNull O observer, @NonNull SCMHeadEvent<?> event,
+                                                     @CheckForNull TaskListener listener)
+            throws IOException, InterruptedException {
+        retrieve(criteria, observer, event, defaultListener(listener));
+        return observer;
+    }
+
+    /**
+     * Fetches the latest heads and corresponding revisions scoped against a specific event.
+     * Implementers are free to cache intermediary results but the call must always check the validity of any
+     * intermediary caches. Where the event {@link SCMEvent#isPayloadTrusted()}, implementations are free to use
+     * the event information to check the validity of any intermediary caches..
+     *
+     * @param <O> Observer type.
+     * @param observer an optional observer of interim results.
+     * @param event the event from which the fetch should be scoped.
+     * @param listener the task listener
+     * @return the provided observer.
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     * @see SCMEvent#isPayloadTrusted()
+     * @since 2.0
+     */
+    @NonNull
+    public final <O extends SCMHeadObserver> O fetch(@NonNull O observer, @NonNull SCMHeadEvent<?> event,
+                                                     @CheckForNull TaskListener listener)
+            throws IOException, InterruptedException {
+        retrieve(getCriteria(), observer, event, defaultListener(listener));
+        return observer;
+    }
+
+    /**
+     * Fetches the latest heads and corresponding revisions. Implementers are free to cache intermediary results
+     * but the call must always check the validity of any intermediary caches.
+     *
+     * @param observer an optional observer of interim results.
+     * @param listener the task listener.
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     */
+    protected void retrieve(@NonNull SCMHeadObserver observer, @NonNull TaskListener listener)
```

Comment:
Especially since they seem to call each other back and forth without real distinction.